--- a/drivers/net/wireless/ath/ath9k/init.c
+++ b/drivers/net/wireless/ath/ath9k/init.c
@@ -97,6 +97,13 @@ static const struct ieee80211_tpt_blink
 };
 #endif
 
+struct ieee80211_hw_txpower_range ath9k_txpwr_range = {
+	.start_idx = 0,
+	.start_pwr = 0,
+	.n_levels = 64,
+	.pwr_step = 2
+};
+
 static int __init set_use_msi(const struct dmi_system_id *dmi)
 {
 	ath9k_use_msi = 1;
@@ -974,6 +981,7 @@ static void ath9k_set_hw_capab(struct at
 	ieee80211_hw_set(hw, SUPPORT_FAST_XMIT);
 	ieee80211_hw_set(hw, SUPPORTS_CLONED_SKBS);
 	ieee80211_hw_set(hw, MFP_CAPABLE);
+	ieee80211_hw_set(hw, SUPPORTS_TPC_PER_MRR);
 
 	if (ath9k_ps_enable)
 		ieee80211_hw_set(hw, SUPPORTS_PS);
@@ -1056,6 +1064,9 @@ static void ath9k_set_hw_capab(struct at
 	wiphy_ext_feature_set(hw->wiphy,
 			      NL80211_EXT_FEATURE_MULTICAST_REGISTRATIONS);
 	wiphy_ext_feature_set(hw->wiphy, NL80211_EXT_FEATURE_CAN_REPLACE_PTK0);
+
+	hw->txpower_ranges = &ath9k_txpwr_range;
+	hw->n_txpower_ranges = 1;
 }
 
 static void ath_get_initial_entropy(struct ath_softc *sc)
